modules/axtask/src/run_queue.rs

```diff
@@ -275,18 +290,33 @@ impl<'a, G: BaseGuard> CurrentRunQueueRef<'a, G> {
     }
 
     /// Migrate the current task to a new run queue matching its CPU affinity and reschedule.
-    /// This function will put the current task into a **new** run queue with `Ready` state,
+    /// This function will put the current task into per CPU variable [`MIGRATING_TASK`] with `Ready` state,
     /// and reschedule to the next task on **this** run queue.
+    ///
+    /// Note: the migrating task stored in [`MIGRATING_TASK`] will be put into the target run queue
+    /// in `finish_task_switch()` after the context switch is completed.
+    #[cfg(feature = "smp")]
     pub fn migrate_current(&mut self) {
-        let curr = &self.current_task;
+        let curr = self.current_task.clone();
         trace!("task migrate: {}", curr.id_name());
         assert!(curr.is_running());
 
-        select_run_queue::<NoOp>(curr.as_task_ref())
-            .inner
-            .put_task_with_state(curr.clone(), TaskState::Running, false);
-
-        self.inner.resched();
+        const MIGRATION_TASK_STACK_SIZE: usize = 4096;
+
+        // Do not put current task to the run queue,
+        // instead, spawn a new task for migrating.
+        let migration = TaskInner::new(
+            move || {
+                curr.set_state(TaskState::Ready);
+                select_run_queue::<kernel_guard::NoPreemptIrqSave>(&curr).add_task(curr);
+            },
+            "migration-task".into(),
+            MIGRATION_TASK_STACK_SIZE,
+        )
+        .into_arc();
+
+        // Call `switch_to` to reschedule to the migration task that performs the migration directly.
+        self.inner.switch_to(crate::current(), migration);
```

Comment:
why not use `curr`?

Reply:
the first argument of `switch_to` is `CurrentTask`, which does not implement `Copy`